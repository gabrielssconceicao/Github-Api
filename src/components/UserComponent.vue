<script setup lang="ts">
import type { IUser } from '@/types/user';
import type { RouterLink } from 'vue-router';
const props = defineProps<{ user: IUser }>()
</script>
<template>
  <section class="h-[calc(100dvh-149.5px)] md:h-[calc(100vh-88px)] bg-section flex justify-center items-center p-4  ">
    <div class="
      container max-w-[500px]
      flex flex-col items-center justify-between gap-3
    bg-white p-3 rounded-lg shadow-md shadow-[#00000088]
      relative
      after:absolute after:h-2 after:w-full after:bottom-[0] after:rounded-b-lg after:bg-primary-contrast ">
      <div class="flex flex-col items-center justify-between gap-3" id="user-info
      ">
        <img :src="props.user.avatar_url" :alt="`${props.user.login} avatar`"
          class="rounded-full w-28 border-4 border-primary" aria-label="avatar" />
        <h2 v-if="props.user.name" class="font-bold font-sans text-xl tracking-wider" aria-label="user-name">{{
          props.user.name }}
        </h2>
        <a class="font-mono font-semibold text-lg bg-primary-contrast flex items-center gap-2 p-1 rounded-md"
          :href="props.user.html_url" target="_blank">
          <img src="@/assets/images/github-logo.svg" alt="github-logo" class="w-6" aria-label="github-logo">
          <span class="text-white" aria-label="github-url">{{ props.user.login }}</span></a>
        <p v-if="props.user.bio" class="font-mono break-before-auto" aria-label="bio">{{ props.user.bio }}</p>
      </div>

      <div class="user-stats flex flex-col gap-3 w-[50%]">
        <div class="follows flex justify-around gap-2 flex-1">
          <p class=" stat" aria-label="followers">Followers <span>{{ props.user.followers }}</span></p>
          <p class="stat" aria-label="following">Following <span>{{ props.user.following }}</span></p>
        </div>
        <div class="flex justify-around gap-2 flex-1">
        </div>
        <p class="stat flex-1" v-if="props.user.location" aria-label="location">Location
          <span>{{ props.user.location }}</span>
        </p>
      </div>

      <RouterLink to="/repos"
        class="bg-primary p-2 w-full text-center text-white hover:bg-primary-contrast font-mono rounded-md">View
        Repositories</RouterLink>

    </div>


  </section>
</template>
