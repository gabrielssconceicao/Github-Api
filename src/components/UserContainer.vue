<script setup lang="ts"></script>
<template>
  <section class="h-[calc(100dvh-149.5px)] md:h-[calc(100vh-88px)] bg-section flex justify-center items-center p-4">
    <slot></slot>
  </section>
</template>
