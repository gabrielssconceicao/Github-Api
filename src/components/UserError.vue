<script setup lang="ts"></script>
<template>
  <section
    class="flex flex-col items-center justify-center flex-1 gap-3 max-w-[400px] min-h-60 bg-white rounded-lg font-mono px-1">
    <h2 class="text-3xl font-bold  text-red-600 mb-4">User Not Found
    </h2>
    <p class="text-gray-800 font-semibold text-center">Sorry, we couldn't find the user you were looking for.</p>
  </section>
</template>
<style>
h2 {
  text-shadow: 2px 1px 2px #00000088;
}
</style>
