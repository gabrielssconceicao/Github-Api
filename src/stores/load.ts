import { ref } from 'vue';
import { defineStore } from 'pinia';

export const useLoadStore = defineStore('load', () => {
  const loading = ref(false);
  function setLoading() {
    loading.value = true;
  }
  function setNotLoading() {
    loading.value = false;
  }
  return { loading, setLoading, setNotLoading };
});
