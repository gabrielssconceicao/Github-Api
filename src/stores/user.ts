import { ref } from 'vue';
import { defineStore } from 'pinia';
import type { IUserResponse } from '@/types/user';
import { fetchUser } from '@/hooks/fetchUser';
export const useUserStore = defineStore('counter', () => {
  const githubUser = ref<IUserResponse>({} as IUserResponse);

  async function searchUser(userLogin: string) {
    const data = await fetchUser(userLogin);
    githubUser.value = data;
  }

  return { githubUser, searchUser };
});
