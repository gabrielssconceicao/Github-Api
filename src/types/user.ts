export interface IUser {
  name: null | string;
  login: string;
  avatar_url: string;
  followers: number;
  following: number;
  location: null | string;
  html_url: string;
  bio: null | string;
  public_repos: number;
  repos_url: string;
}

export interface IUserResponse {
  user: IUser;
  status: number;
}
