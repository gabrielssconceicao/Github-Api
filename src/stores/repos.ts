import { defineStore } from 'pinia';
import { ref, computed } from 'vue';

import { type IRepos } from '@/types/repos';

import { langSeparator } from '@/helpers/langSeparator';
import { useLoadStore } from './load';
export const useReposStore = defineStore('repos', () => {
  const repos = ref<IRepos[]>([]);
  const reposUrl = ref('');
  const filterValue = ref('');

  function setRepos(reposValue: IRepos[]) {
    repos.value = reposValue;
  }

  const setFilter = (value: string) => {
    filterValue.value = value;
  };

  function setReposUrl(url: string) {
    reposUrl.value = url;
  }
  async function getRepos() {
    const useLoad = useLoadStore();
    useLoad.setLoading();
    try {
      const response = await fetch(reposUrl.value);
      const data: IRepos[] = await response.json();
      const dataWithoutNull = data.map((e) => {
        if (e.language) return e;
        return { ...e, language: 'No Language' };
      });
      setRepos(dataWithoutNull);
    } catch (error) {
      setRepos([]);
    } finally {
      useLoad.setNotLoading();
    }
  }

  const filteredRepos = computed(() => {
    return filterValue.value !== ''
      ? repos.value.filter((repo) => repo.language === filterValue.value)
      : repos.value;
  });

  const filteredReposCount = computed(
    () => repos.value && langSeparator(repos.value),
  );

  return {
    filteredRepos,
    getRepos,
    filteredReposCount,
    setFilter,
    setReposUrl,
    setRepos,
    repos,
    reposUrl,
    filterValue,
  };
});
