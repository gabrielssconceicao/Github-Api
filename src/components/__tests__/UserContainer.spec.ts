import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';

import UserContainer from '../UserContainer.vue';
import UserComponent from '../UserComponent.vue';
import UserError from '../UserError.vue';
import Loading from '../LoadingComponent.vue';

import { useUserStore } from '@/stores/user';

describe('<UserContainer />', () => {
  it('should render user container', () => {
    const wrapper = mount(UserContainer);
    expect(wrapper.find('section').exists()).toBe(true);

    expect(wrapper.findComponent(UserComponent).exists()).toBe(false);
    expect(wrapper.findComponent(UserError).exists()).toBe(false);
    expect(wrapper.findComponent(Loading).exists()).toBe(false);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should show loadComponent when there is a search', async () => {
    const wrapper = mount(UserContainer);
    const userStore = useUserStore();
    userStore.setInputValue('resolveTest');
    userStore.searchUser();
    await wrapper.vm.$nextTick();
    expect(wrapper.findComponent(UserError).exists()).toBe(false);
    expect(wrapper.findComponent(UserComponent).exists()).toBe(false);
    expect(wrapper.findComponent(Loading).exists()).toBe(true);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should show UserComponent when there is a successful search', async () => {
    const wrapper = mount(UserContainer);
    const userStore = useUserStore();
    userStore.setInputValue('resolveTest');
    await userStore.searchUser();

    expect(wrapper.findComponent(Loading).exists()).toBe(false);
    expect(wrapper.findComponent(UserError).exists()).toBe(false);
    expect(wrapper.findComponent(UserComponent).exists()).toBe(true);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should show Error when there is a failed search', async () => {
    const wrapper = mount(UserContainer);
    const userStore = useUserStore();
    userStore.setInputValue('rejectTest');
    await userStore.searchUser();

    expect(wrapper.findComponent(Loading).exists()).toBe(false);
    expect(wrapper.findComponent(UserComponent).exists()).toBe(false);
    expect(wrapper.findComponent(UserError).exists()).toBe(true);
    expect(wrapper.element).toMatchSnapshot();
  });
});
