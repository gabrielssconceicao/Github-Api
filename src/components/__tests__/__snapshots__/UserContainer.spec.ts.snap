// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<UserContainer /> > should render user container 1`] = `
<section
  class="min-h-[calc(100dvh-149.5px)] md:min-h-[calc(100vh-88px)] bg-section flex justify-center items-center p-4"
>
  
  
</section>
`;

exports[`<UserContainer /> > should render with slots 1`] = `
<section
  class="min-h-[calc(100dvh-149.5px)] md:min-h-[calc(100vh-88px)] bg-section flex justify-center items-center p-4"
>
  
  <h2>
    Test
  </h2>
  
</section>
`;
